specifyweb/frontend/js_src/lib/components/AttachmentsBulkImport/Import.tsx

```diff
@@ -0,0 +1,368 @@
+                `/attachment_gw/dataset/${resourceId}/${id}/`,
+                {
+                  headers: { Accept: 'application/json' },
+                  method: 'GET',
+                }
+              ).then(async ({ data }) => {
+                if (data.status === 'main') return data;
+                const reconstructFunction =
+                  data.status === 'uploading'
+                    ? reconstructUploadingAttachmentSpec
+                    : reconstructDeletingAttachment;
+                return reconstructFunction(
+                  data.uploadSpec.staticPathKey,
+                  data.uploadableFiles
+                ).then((returnFiles) => ({
+                  ...data,
+                  uploadableFiles: returnFiles,
+                }));
+              })
+        ),
+      [id]
+    ),
+    true
+  );
+  return attachmentDataSet === undefined ? null : (
+    <AttachmentsImport attachmentDataSetResource={attachmentDataSet} />
+  );
+}
+
+function AttachmentsImport<DATASET extends AttachmentDataSet>({
+  attachmentDataSetResource,
+}: {
+  readonly attachmentDataSetResource: DATASET;
+}): JSX.Element | null {
+  const [eagerDataSet, isSaving, isBrandNew, triggerSave, commitChange] =
+    useEagerDataSet(attachmentDataSetResource);
+
+  const commitFileChange = (
+    newUploadables: (
+      oldUploadables: RA<PartialUploadableFileSpec>
+    ) => RA<PartialUploadableFileSpec>
+  ): void =>
+    commitChange((oldState) => ({
+      ...oldState,
+      uploadableFiles: newUploadables(oldState.uploadableFiles),
+    }));
+
+  const commitStatusChange = (newState: AttachmentDataSet['status']) =>
+    commitChange((oldState) => ({ ...oldState, status: newState }));
+
+  const applyFileNames = React.useCallback(
+    (file: UnBoundFile): PartialUploadableFileSpec =>
+      eagerDataSet.uploadSpec.staticPathKey === undefined
+        ? { file }
+        : resolveFileNames(
+            file,
+            eagerDataSet.uploadSpec.formatQueryResults,
+            eagerDataSet.uploadSpec.fieldFormatter
+          ),
+    [eagerDataSet.uploadSpec.staticPathKey]
+  );
+
+  const previousKeyRef = React.useRef(
+    attachmentDataSetResource.uploadSpec.staticPathKey
+  );
+  React.useEffect(() => {
+    // Reset all parsed names if matching path is changes
+    if (previousKeyRef.current !== eagerDataSet.uploadSpec.staticPathKey) {
+      previousKeyRef.current = eagerDataSet.uploadSpec.staticPathKey;
+      commitFileChange((files) =>
+        files.map((file) => applyFileNames(file.file))
+      );
+    }
+  }, [applyFileNames, commitFileChange]);
+
+  const currentBaseTable =
+    eagerDataSet.uploadSpec.staticPathKey === undefined
+      ? undefined
+      : staticAttachmentImportPaths[eagerDataSet.uploadSpec.staticPathKey]
+          .baseTable;
+
+  const anyUploaded = eagerDataSet.uploadableFiles.some(
+    (uploadable) => uploadable.attachmentId !== undefined
+  );
+  const handleFilesSelected = React.useCallback(
```

Comment:
It doesn't need to be one! I personally like using memoization because it makes bugs easier to catch. For example, I expect this function to only change meaningfully with `applyFileNames` and `commitChange`, and only change its behaviour in those cases. However, if some other component / hook expected to change with some other state, it would appear more easily as a bug. 

Memoization does have it's downside (consumes memory). But in this case, it can be changed to not have memoization, because as you said, no dependencies. Thanks for catching this!

Reply:
huh, I don't get the explanation 😵‍💫
"I personally like using memoization because it makes bugs easier to catch"
isn't memoization the very reason this bug exists in the first place?

if it's not memoizated, then you always have the freshest version of the function and thus you don't have an issue with stale values.
(of course, you have to memorize in react when you use a function in the dependency array)

Libraries that were designed after react have actually made this less of an issue than react - in react, creating a ref is frowned upon because ref is outside the reactive system (doesn't cause re-render), but you get the most up to date value any were you check the ref's value. In other libraries, creating a ref is very easy (or is even the default behavior), and you only create a state when you actually want reactivity

Reply:
> isn't memoization the very reason this bug exists in the first place?

But there isn't a bug right now!

> if it's not memoizated, then you always have the freshest version of the function and thus you don't have an issue with stale values.

But if the behaviour of function should only be dependent on some states, why not memoize?